### DISCUSSION TITLE
Tornado Mobile click events not detected with socketIO

### DISCUSSION THREAD

**Discussion by csharmatech** [miguelgrinberg/python-socketio#1102] [Q&A] [answered]
Mobile click events are not detected when using Tornado + SocketIO. It works fine on laptop browser.

**Server Code:**

```
import os
import sys
import socketio
import subprocess
import tornado.ioloop
from tornado.options import define, options

sio = socketio.AsyncServer(cors_allowed_origins='*',async_mode='tornado')

async def ssh():
    cmd1 = 'systemctl status sshd| grep Active | awk -F " " \'{print$2}\''
    listing1 = subprocess.run(cmd1,stdout=subprocess.PIPE,shell=True,universal_newlines=True)
    result1 = listing1.stdout
    await asyncio.sleep(5)
    return result1

@sio.on('my_event')
async def servermessage(data):
    data = asyncio.create_task(ssh())
    while True:
        if data.done():
            value = data.result()
            await sio.emit('response', "SSH Status is:" + " " + value)
            break
        else:
            await sio.emit('response', "Wait..")

def make_app():
    return tornado.web.Application([
        (r"/", MainHandler),
        (r"/socket.io/", socketio.get_tornado_handler(sio)),
        
    ],template_path='/home/test/tornado/templates',debug=True)

class MainHandler(tornado.web.RequestHandler):
    def get(self):
        self.render("base.html")

settings = dict(
        template_path=os.path.join(os.path.dirname(__file__), "templates"),
)

async def main():
    app = make_app()
    app.listen(8888)
    shutdown_event = asyncio.Event()
    await shutdown_event.wait()
    tornado.ioloop.IOLoop.current().start()

if __name__ == "__main__":
    asyncio.run(main())
```

**Client side Javascript code**

```
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta http-equiv="Cache-control" content="no-cache" charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
<script type="text/javascript" src="//code.jquery.com/jquery-1.4.2.min.js"></script>
<script type="text/javascript" src="//cdn.socket.io/4.4.1/socket.io.min.js"></script>
<link href="https://cdn.jsdelivr.net/npm/bootstrap@5.2.3/dist/css/bootstrap.min.css" rel="stylesheet">
<script src="https://cdn.jsdelivr.net/npm/bootstrap@5.2.3/dist/js/bootstrap.bundle.min.js"></script>
<link href='http://fonts.googleapis.com/css?family=Oswald:400,300,700' rel='stylesheet' type='text/css'>
<script type="text/javascript" charset="utf-8">
const socket = io("http://<IP>:8888");
socket.on('connect', function (data) {
  console.log('Successfully connected!');
});
socket.on('response', function(data) {
document.getElementById("test").innerHTML=data;

});

function myupdate(event) {
	socket.emit("my_event", function(msg) {
       });
}
</script>
</head>
<style>
</style>
<body>
<center>
<h1 style="color:black;background-color:#33ACFF;padding:30px">Welcome to Tornado</h1>
</center>
<button class="btn btn-info" id="button1" style="margin:5px" onclick="myupdate()">Check Services</button>
<p id ="test"></p>
</body>
```

**Comment by csharmatech** [comment]
![image](https://user-images.githubusercontent.com/85521686/208254508-4887abcd-6a77-4721-8d3a-d10186a78108.png)

Above is the capture from Laptop browser which is expected

**Comment by csharmatech** [comment]
Below is the screenshot from mobile browser where nothing happens after clicking button. It seems like button click is not detected using client emit ??

![image](https://user-images.githubusercontent.com/85521686/208254605-dd098bfc-4bc2-46b2-b222-e27ae6c4335a.png)

**Comment by miguelgrinberg** [comment, marked as answer]
Please review the information in the Troubleshooting section of the documentation. This should help you debug this problem, which to me seems to be unrelated to Socket.IO. https://python-socketio.readthedocs.io/en/latest/server.html#debugging-and-troubleshooting

**Reply by csharmatech** [reply]
Hi @miguelgrinberg ,

I enabled the logging, however not seems useful as it prints nothing as click is not sending any events. This has been issue for frameworks I am exploring - FastAPI, Tornado

The same code works fine on Flask.

Not sure why mobile click events are not being detected? If I remove the socket.emit and replace with a simple Javascript alert, the button click works fine. So it seems like issue with socket.emit, the mobile click event is not detected??

**Reply by miguelgrinberg** [reply]
I really don't know. This 'socket.emit` that is at the center of your issue is not part of this project, you are using the JavaScript Socket.IO client for that. Aside from debugging your front end code I don't have anything else to suggest, sorry.

> The same code works fine on Flask.

That cannot be. The server code that you are using is asynchronous, that code would not work with Flask. I also fail to see what Flask vs. Tornado have to do with the issue if the events are not being sent from the client.

**Reply by csharmatech** [reply]
I am sorry for my above comment on Flask.

I meant that,  I used threading in Flask and it works fine with socketIO and mobile button click works fine.

**Reply by csharmatech** [reply]
I am trying to explore the asynchronous frameworks as I need to use asyncio feature and use my above code on mobile.

**Reply by csharmatech** [reply]
@miguelgrinberg , Thank you for your quick responses and feedback. I will try to debug more on the mobile part related issues.

**Reply by csharmatech** [reply]
Finally got it working..

So firstly what I eventually found out was that the websocket connection never got opened on mobile. Although the page got displayed, but the connection was not open and that's the reason that button click never worked. Secondly, I had to add below line to finally make the connection open on mobile. My webserver is running in KVM environment with port forwarding.

```
$(document).ready(function() {  // This line was missing
const socket = io("http://<IP>:8888/");
});
```
